azurerm/resource_arm_firewall_application_rule_collection.go

```diff
@@ -0,0 +1,448 @@
+			"action": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(network.AzureFirewallRCActionTypeAllow),
+					string(network.AzureFirewallRCActionTypeDeny),
+				}, false),
+			},
+
+			"rule": {
+				Type:     schema.TypeList,
+				Required: true,
+				MinItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.NoZeroValues,
+						},
+						"description": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"source_addresses": {
+							Type:     schema.TypeSet,
+							Required: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Set:      schema.HashString,
+						},
+						"fqdn_tags": {
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Set:      schema.HashString,
+						},
+						"target_fqdns": {
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Set:      schema.HashString,
+						},
+						"protocol": {
+							Type:     schema.TypeList,
+							Optional: true,
+							MinItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"port": {
+										Type:         schema.TypeInt,
+										Optional:     true,
+										ValidateFunc: validate.PortNumber,
+									},
+									"type": {
+										Type:     schema.TypeString,
+										Required: true,
+										ValidateFunc: validation.StringInSlice([]string{
+											string(network.AzureFirewallApplicationRuleProtocolTypeHTTP),
+											string(network.AzureFirewallApplicationRuleProtocolTypeHTTPS),
+										}, false),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceArmFirewallApplicationRuleCollectionCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).azureFirewallsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name := d.Get("name").(string)
+	firewallName := d.Get("azure_firewall_name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+	applicationRules, err := expandArmFirewallApplicationRules(d.Get("rule").([]interface{}))
+	if err != nil {
+		return fmt.Errorf("Error expanding Firewall Application Rules: %+v", err)
+	}
+
+	azureRMLockByName(firewallName, azureFirewallResourceName)
+	defer azureRMUnlockByName(firewallName, azureFirewallResourceName)
+
+	firewall, err := client.Get(ctx, resourceGroup, firewallName)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Firewall %q (Resource Group %q): %+v", firewallName, resourceGroup, err)
+	}
+
+	if firewall.AzureFirewallPropertiesFormat == nil {
+		return fmt.Errorf("Error expanding Firewall %q (Resource Group %q): `properties` was nil", firewallName, resourceGroup)
```

Comment:
Would this read better as:
```suggestion
		return fmt.Errorf("Unable to parse Firewall %q (Resource Group %q): `properties` was nil", firewallName, resourceGroup)
```

Reply:
We are not in an expand function here so I feel like a different term should be used. But it's rather minor and not a big deal 🙂 I'll leave it up to you.

Reply:
I went with `Error retrieving` then it matches the Read function